internal/contour/cluster.go

```diff
@@ -15,77 +15,37 @@ package contour
 
 import (
 	"sort"
-	"sync"
-
-	resource "github.com/envoyproxy/go-control-plane/pkg/resource/v2"
 
 	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
-	"github.com/golang/protobuf/proto"
+	xds "github.com/envoyproxy/go-control-plane/pkg/cache/types"
 	"github.com/projectcontour/contour/internal/dag"
 	"github.com/projectcontour/contour/internal/envoy"
 	"github.com/projectcontour/contour/internal/protobuf"
 	"github.com/projectcontour/contour/internal/sorter"
 )
 
-// ClusterCache manages the contents of the gRPC CDS cache.
-type ClusterCache struct {
-	mu     sync.Mutex
-	values map[string]*v2.Cluster
-	Cond
-}
-
-// Update replaces the contents of the cache with the supplied map.
-func (c *ClusterCache) Update(v map[string]*v2.Cluster) {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-
-	c.values = v
-	c.Cond.Notify()
-}
-
-// Contents returns a copy of the cache's contents.
-func (c *ClusterCache) Contents() []proto.Message {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+// translateClusters returns an array of CDS resources.
+func translateClusters(clusters map[string]*v2.Cluster) []xds.Resource {
 	var values []*v2.Cluster
-	for _, v := range c.values {
+	for _, v := range clusters {
 		values = append(values, v)
 	}
```

Comment:
We could avoid having to do this iteration to convert from a map to a slice, if we change `*clusterVisitor.visit(...)` to produce a slice instead of a map. There's not actually any reason to produce the map anymore, since we never need to do a lookup by key on the end result (previously we did though). It'd be slightly more efficient if we switched to just a slice, but probably doesn't matter too much in practice. Something to keep in mind.

The implementation of `*clusterVisitor.visit(...)` _would_ still need to keep a local map of cluster names it's seen already, to avoid putting dupes in the slice it's producing, but that could be just a local var separate from the slice being produced.